lib/Rex/Resource/file/Provider/base.pm

```diff
@@ -76,7 +76,10 @@ sub test {
   my $is_dir    = $self->is_dir;
   my $is_absent = any { $self->config->{ensure} eq $_ } qw/absent/;
 
-  if ( $is_file && $self->fs->is_file($file) ) {
+  if ( $is_file
+    && $self->fs->is_file($file)
+    && ( $self->config->{source} || $self->config->{content} ) )
```

Comment:
I can't comment directly on the line below, but I think `$self->_set_upload(0);` still should be called even if the checksum calculation is skipped, to signal "there's no need to change the _content_ of the file" (but other changes might still be necessary).